testnet-contracts/contracts/Peggy.sol

```diff
@@ -1,18 +1,20 @@
 pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
 
 import "./Processor.sol";
+import "./CosmosBridge.sol";
+import "./Oracle.sol";
+import "./Bank.sol";
 
   /*
    *  @title: Peggy
-   *  @dev: Peg zone contract for testing one-way transfers from Ethereum
-   *        to Cosmos, facilitated by a trusted provider. This contract is
-   *        NOT intended to be used in production and users are empowered
-   *        to withdraw their locked funds at any time.
+   *  @dev: Peg zone contract two-way asset transfers between Ethereum and
+   *        to Cosmos, facilitated by a set of validators. This contract is
+   *        NOT intended to be used in production (yet).
    */
-contract Peggy is Processor {
+contract Peggy is CosmosBridge, Oracle, Bank, Processor {
```

Comment:
I think we may want to align the solidity implementation with the go implementation. Specifically what comes to mind is that in the Go implementation we have something like:
 - transaction first goes to bridge module
 - bridge module *uses* oracle module
 - oracle module returns result to bridge module
 - bridge module processes based on result

compared to this implementation, where we have a Peggy contract that is inheriting all the above functionality and state and is this single massive context doing everything.

Are there any issues/concerns/gas inefficiencies with the componentization/code abstraction based on the Go version that are not applicable or non-standard in Solidity, or would the Go abstractions work fine? Can discuss on call.

Reply:
As of v0.5.0, "Specifying base constructor arguments multiple times in the same inheritance hierarchy is now disallowed". Ideally the Bridge contract inherits from the Oracle contract which inherits from a Valset contract. However, since our Valset is deployed with constructor arguments `initValidatorAddresses` and `initValidatorPowers` we can only pass them once through the inheritance hierarchy. Hence the subsequent decision to combine Valset + Oracle and have Peggy inherit from it directly.

There are multiple other potential solutions (no constructor arguments just call ```updateValidatorsPower(newValidators, newPowers)```, or deploy the Valset as an isolated contract, for example) which we can discuss.